File: sdk/typescript/tests/integration.test.ts
/**
 * Prismer TypeScript SDK — Comprehensive Integration Tests
 *
 * Runs against the live production environment (https://prismer.cloud).
 * Requires PRISMER_API_KEY_TEST env var.
 *
 * Usage:
 *   PRISMER_API_KEY_TEST="sk-prismer-live-..." npx vitest run tests/integration.test.ts --reporter=verbose
 */

import { describe, it, expect, beforeAll } from 'vitest';
import { PrismerClient } from '../src/index';

// Increase default test timeout for integration tests hitting a live API.
// Individual slow tests (search, PDF parse) get even longer timeouts below.

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

const API_KEY = process.env.PRISMER_API_KEY_TEST;
if (!API_KEY) {
  throw new Error('PRISMER_API_KEY_TEST environment variable is required');
}

const RUN_ID = Date.now().toString(36); // unique per run to avoid collisions

/** Create a client authenticated with the API key */
function apiClient(): PrismerClient {
  return new PrismerClient({
    apiKey: API_KEY!,
    environment: 'production',
    timeout: 60_000,
  });
}

/** Create a client authenticated with an IM JWT token */
function imClient(token: string): PrismerClient {
  return new PrismerClient({
    apiKey: token,
    environment: 'production',
    timeout: 60_000,
  });
}

// ---------------------------------------------------------------------------
// Group 1: Context API
// ---------------------------------------------------------------------------

describe('Context API', () => {
  const client = apiClient();

  it('load() single URL — returns success with mode single_url', async () => {
    const result = await client.load('https://example.com');
    expect(result.success).toBe(true);
    expect(result.mode).toBe('single_url');
    expect(result.result).toBeDefined();
    expect(result.result!.url).toContain('example.com');
    // hqcc may or may not exist depending on cache state
    expect(typeof result.result!.cached).toBe('boolean');
  });

  it('load() batch URLs — returns success with mode batch_urls', async () => {
    const result = await client.load([
      'https://example.com',
      'https://httpbin.org/html',
    ]);
    expect(result.success).toBe(true);
    expect(result.mode).toBe('batch_urls');
    expect(result.results).toBeDefined();
    expect(Array.isArray(result.results)).toBe(true);
    expect(result.results!.length).toBeGreaterThanOrEqual(1);
  });

  it('load() search query — returns success', async () => {
    const result = await client.load('What is TypeScript?', {
      inputType: 'query',
    });
    expect(result.success).toBe(true);
    // mode could be 'query' for search-based load
    expect(result.requestId).toBeDefined();
  }, 60_000);

  it('save() — saves content and returns success', async () => {
    const result = await client.save({
      url: `https://test-${RUN_ID}.example.com/integration-test`,
      hqcc: `# Integration Test Content\n\nSaved at ${new Date().toISOString()} by run ${RUN_ID}.`,
    });
    expect(result.success).toBe(true);
  });

  it('search() — performs a search query', async () => {
    const result = await client.search('example domain');
    expect(result.success).toBe(true);
    expect(result.requestId).toBeDefined();
  }, 60_000);
});

// ---------------------------------------------------------------------------
// Group 2: Parse API
// ---------------------------------------------------------------------------

describe('Parse API', () => {
  const client = apiClient();

  it('parsePdf() with URL — returns success and requestId', async () => {
    const result = await client.parsePdf(
      'https://arxiv.org/pdf/2401.00001.pdf',
      'fast',
    );
    expect(result.success).toBe(true);
    expect(result.requestId).toBeDefined();
    // The response may be synchronous (document) or async (taskId)
    const hasDocOrTask = result.document !== undefined || result.taskId !== undefined;
    expect(hasDocOrTask).toBe(true);
  }, 60_000);

  it('parse() with mode auto — returns success', async () => {
    const result = await client.parse({
      url: 'https://arxiv.org/pdf/2401.00001.pdf',
      mode: 'auto',
    });
    expect(result.success).toBe(true);
    expect(result.requestId).toBeDefined();
  }, 60_000);
});

// ---------------------------------------------------------------------------
// Group 3: IM API — Full Lifecycle
// ---------------------------------------------------------------------------

describe('IM API', () => {
  const client = apiClient();

  // State shared across IM tests
  let agentAToken: string;
  let agentAId: string;
  let agentAUsername: string;
  let agentBToken: string;
  let agentBId: string;
  let agentBUsername: string;
  let clientA: PrismerClient; // authenticated with agent A's JWT
  let clientB: PrismerClient; // authenticated with agent B's JWT
  let directConversationId: string;
  let groupId: string;

  // -----------------------------------------------------------------------
  // Account
  // -----------------------------------------------------------------------

  describe('Account', () => {
    it('register() agent A — returns isNew=true and token', async () => {
      agentAUsername = `test-agent-a-${RUN_ID}`;
      const reg = await client.im.account.register({
        type: 'agent',
        username: agentAUsername,
        displayName: `Test Agent A (${RUN_ID})`,
        agentType: 'assistant',
        capabilities: ['testing', 'integration'],
        description: 'Integration test agent A',
      });
      expect(reg.ok).toBe(true);
      expect(reg.data).toBeDefined();
      expect(reg.data!.isNew).toBe(true);
      expect(reg.data!.token).toBeDefined();
      expect(typeof reg.data!.token).toBe('string');
      expect(reg.data!.imUserId).toBeDefined();

      agentAToken = reg.data!.token;
      agentAId = reg.data!.imUserId;
      clientA = imClient(agentAToken);
    });

    it('register() agent B — second agent as message target', async () => {
      agentBUsername = `test-agent-b-${RUN_ID}`;
      const reg = await client.im.account.register({
        type: 'agent',
        username: agentBUsername,
        displayName: `Test Agent B (${RUN_ID})`,
        agentType: 'specialist',
        capabilities: ['testing'],
        description: 'Integration test agent B',
      });
      expect(reg.ok).toBe(true);
      expect(reg.data).toBeDefined();
      expect(reg.data!.isNew).toBe(true);
      expect(reg.data!.token).toBeDefined();

      agentBToken = reg.data!.token;
      agentBId = reg.data!.imUserId;
      clientB = imClient(agentBToken);
    });

    it('me() — returns user profile and agentCard', async () => {
      const me = await clientA.im.account.me();
      expect(me.ok).toBe(true);
      expect(me.data).toBeDefined();
      expect(me.data!.user).toBeDefined();
      expect(me.data!.user.username).toBe(agentAUsername);
      expect(me.data!.agentCard).toBeDefined();
      expect(me.data!.agentCard!.agentType).toBe('assistant');
    });

    it('refreshToken() — returns a new token', async () => {
      const refresh = await clientA.im.account.refreshToken();
      expect(refresh.ok).toBe(true);
      expect(refresh.data).toBeDefined();
      expect(refresh.data!.token).toBeDefined();
      expect(typeof refresh.data!.token).toBe('string');
      // Update token for subsequent calls
      agentAToken = refresh.data!.token;
      clientA = imClient(agentAToken);
    });
  });

  // -----------------------------------------------------------------------
  // Direct Messaging
  // -----------------------------------------------------------------------

  describe('Direct Messaging', () => {
    it('send() — agent A sends message to agent B', async () => {
      const result = await clientA.im.direct.send(agentBId, 'Hello from Agent A!');
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(result.data!.message).toBeDefined();
      expect(result.data!.message.id).toBeDefined();
      expect(result.data!.conversationId).toBeDefined();
      directConversationId = result.data!.conversationId;
    });

    it('send() — agent B replies to agent A', async () => {
      const result = await clientB.im.direct.send(agentAId, 'Hello back from Agent B!');
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(result.data!.message.id).toBeDefined();
    });

    it('getMessages() — retrieves message history', async () => {
      const result = await clientA.im.direct.getMessages(agentBId);
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(Array.isArray(result.data)).toBe(true);
      expect(result.data!.length).toBeGreaterThanOrEqual(1);
    });
  });

  // -----------------------------------------------------------------------
  // Credits
  // -----------------------------------------------------------------------

  describe('Credits', () => {
    it('get() — returns balance for new agent', async () => {
      const result = await clientA.im.credits.get();
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(typeof result.data!.balance).toBe('number');
    });

    it('transactions() — returns transaction array', async () => {
      const result = await clientA.im.credits.transactions();
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(Array.isArray(result.data)).toBe(true);
    });
  });

  // -----------------------------------------------------------------------
  // Contacts & Discovery
  // -----------------------------------------------------------------------

  describe('Contacts & Discovery', () => {
    it('contacts.list() — returns contacts array', async () => {
      const result = await clientA.im.contacts.list();
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(Array.isArray(result.data)).toBe(true);
      // After messaging agent B, agent B should be in contacts
      if (result.data!.length > 0) {
        const contact = result.data!.find(
          (c) => c.username === agentBUsername,
        );
        expect(contact).toBeDefined();
      }
    });

    it('contacts.discover() — returns array of agents', async () => {
      const result = await clientA.im.contacts.discover();
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(Array.isArray(result.data)).toBe(true);
    });
  });

  // -----------------------------------------------------------------------
  // Groups
  // -----------------------------------------------------------------------

  describe('Groups', () => {
    it('create() — creates a group chat', async () => {
      const result = await clientA.im.groups.create({
        title: `Test Group ${RUN_ID}`,
        description: 'Integration test group',
        members: [agentBId],
      });
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(result.data!.groupId).toBeDefined();
      groupId = result.data!.groupId;
    });

    it('list() — lists groups', async () => {
      const result = await clientA.im.groups.list();
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(Array.isArray(result.data)).toBe(true);
      expect(result.data!.length).toBeGreaterThanOrEqual(1);
    });

    it('get() — gets group details', async () => {
      const result = await clientA.im.groups.get(groupId);
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(result.data!.groupId).toBe(groupId);
      expect(result.data!.title).toContain('Test Group');
    });

    it('send() — sends message to group', async () => {
      const result = await clientA.im.groups.send(groupId, 'Hello group!');
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(result.data!.message).toBeDefined();
    });

    it('addMember() — adds a member (may already be present)', async () => {
      // Agent B is already a member from creation, so this may succeed or return an error.
      // We test the call completes without throwing.
      const result = await clientA.im.groups.addMember(groupId, agentBId);
      // Could be ok=true (added) or ok=false (already member)
      expect(result).toBeDefined();
    });

    it('getMessages() — retrieves group messages', async () => {
      const result = await clientA.im.groups.getMessages(groupId);
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(Array.isArray(result.data)).toBe(true);
      expect(result.data!.length).toBeGreaterThanOrEqual(1);
    });
  });

  // -----------------------------------------------------------------------
  // Conversations
  // -----------------------------------------------------------------------

  describe('Conversations', () => {
    it('list() — returns conversations array', async () => {
      const result = await clientA.im.conversations.list();
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(Array.isArray(result.data)).toBe(true);
      expect(result.data!.length).toBeGreaterThanOrEqual(1);
    });

    it('get() — returns conversation details', async () => {
      expect(directConversationId).toBeDefined();
      const result = await clientA.im.conversations.get(directConversationId);
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(result.data!.id).toBe(directConversationId);
    });

    it('markAsRead() — marks conversation as read', async () => {
      const result = await clientA.im.conversations.markAsRead(directConversationId);
      expect(result.ok).toBe(true);
    });
  });

  // -----------------------------------------------------------------------
  // Messages (low-level by conversationId)
  // -----------------------------------------------------------------------

  describe('Messages (low-level)', () => {
    it('send() — sends message to a conversation', async () => {
      expect(directConversationId).toBeDefined();
      const result = await clientA.im.messages.send(
        directConversationId,
        'Low-level message test',
      );
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(result.data!.message).toBeDefined();
      expect(result.data!.message.content).toBe('Low-level message test');
    });

    it('getHistory() — retrieves messages for conversation', async () => {
      const result = await clientA.im.messages.getHistory(directConversationId);
      expect(result.ok).toBe(true);
      expect(result.data).toBeDefined();
      expect(Array.isArray(result.data)).toBe(true);
      expect(result.data!.length).toBeGreaterThanOrEqual(1);
    });
  });

  // -----------------------------------------------------------------------
  // Workspace
  // -----------------------------------------------------------------------

  describe('Workspace', () => {
    it('init() — initializes a 1:1 workspace', async () => {
      const result = await clientA.im.workspace.init();
      // Workspace may or may not be available in test env
      if (result.ok) {
        expect(result.data).toBeDefined();
        expect(result.data!.workspaceId).toBeDefined();
        expect(result.data!.conversationId).toBeDefined();
      } else {
        // Acceptable: workspace feature may not be enabled
        expect(result.error).toBeDefined();
      }
    });
  });

  // -----------------------------------------------------------------------
  // Edge Cases
  // -----------------------------------------------------------------------

  describe('Edge Cases', () => {
    it('register duplicate username (re-register same agent) — should return isNew=false or error', async () => {
      const reg = await client.im.account.register({
        type: 'agent',
        username: agentAUsername,
        displayName: `Test Agent A duplicate (${RUN_ID})`,
      });
      // Server may return ok with isNew=false (idempotent) or error 409
      if (reg.ok) {
        expect(reg.data!.isNew).toBe(false);
      } else {
        expect(reg.error).toBeDefined();
      }
    });

    it('send to nonexistent user — should fail', async () => {
      const result = await clientA.im.direct.send(
        'nonexistent-user-id-00000000',
        'This should fail',
      );
      expect(result.ok).toBe(false);
      expect(result.error).toBeDefined();
    });

    it('access without auth — should fail (401)', async () => {
      const noAuthClient = new PrismerClient({
        apiKey: 'invalid-token-not-real',
        environment: 'production',
      });
      const result = await noAuthClient.im.account.me();
      expect(result.ok).toBe(false);
      expect(result.error).toBeDefined();
    });
  });
});

File: sdk/golang/integration_test.go
//go:build integration

package prismer_test

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"testing"
	"time"

	prismer "github.com/prismer-io/prismer-sdk-go"
)

// imMessageLoose works around a SDK bug where IMMessage.Metadata is typed as
// map[string]any but the API can return a JSON string for that field.
// See SDK bug report in test summary.
type imMessageLoose struct {
	ID        string          `json:"id"`
	Content   string          `json:"content"`
	Type      string          `json:"type"`
	SenderID  string          `json:"senderId"`
	CreatedAt string          `json:"createdAt"`
	Metadata  json.RawMessage `json:"metadata,omitempty"`
}

// helpers ---------------------------------------------------------------

func apiKey(t *testing.T) string {
	t.Helper()
	key := os.Getenv("PRISMER_API_KEY_TEST")
	if key == "" {
		t.Fatal("PRISMER_API_KEY_TEST environment variable is required")
	}
	return key
}

func newClient(t *testing.T) *prismer.Client {
	t.Helper()
	return prismer.NewClient(apiKey(t), prismer.WithEnvironment(prismer.Production))
}

func uniqueName(prefix string) string {
	return fmt.Sprintf("%s_%d", prefix, time.Now().UnixNano())
}

// =======================================================================
// Group 1: Context API
// =======================================================================

func TestIntegration_Context_LoadSingle(t *testing.T) {
	client := newClient(t)
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	result, err := client.Load(ctx, "https://example.com", nil)
	if err != nil {
		t.Fatalf("Load returned error: %v", err)
	}
	if !result.Success {
		t.Fatalf("Load was not successful: %+v", result.Error)
	}
	if result.Mode == "" {
		t.Error("expected non-empty Mode")
	}
	t.Logf("Load single — mode=%s requestId=%s", result.Mode, result.RequestID)

	// In single mode we expect Result to be populated
	if result.Result == nil && len(result.Results) == 0 {
		t.Error("expected at least Result or Results to be populated")
	}
}

func TestIntegration_Context_LoadBatch(t *testing.T) {
	client := newClient(t)
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	urls := []string{"https://example.com", "https://httpbin.org/html"}
	result, err := client.Load(ctx, urls, nil)
	if err != nil {
		t.Fatalf("Load batch returned error: %v", err)
	}
	if !result.Success {
		t.Fatalf("Load batch was not successful: %+v", result.Error)
	}
	t.Logf("Load batch — mode=%s results=%d", result.Mode, len(result.Results))

	if result.Mode != "batch_urls" {
		t.Errorf("expected mode=batch_urls, got %s", result.Mode)
	}
	if len(result.Results) == 0 {
		t.Error("expected non-empty Results for batch load")
	}
}

func TestIntegration_Context_Save(t *testing.T) {
	client := newClient(t)
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	saveURL := fmt.Sprintf("https://test.example.com/go-integration-%d", time.Now().UnixNano())
	result, err := client.Save(ctx, &prismer.SaveOptions{
		URL:  saveURL,
		HQCC: "Integration test content from Go SDK.",
	})
	if err != nil {
		t.Fatalf("Save returned error: %v", err)
	}
	if !result.Success {
		t.Fatalf("Save was not successful: %+v", result.Error)
	}
	t.Logf("Save — status=%s url=%s", result.Status, result.URL)
}

// =======================================================================
// Group 2: Parse API
// =======================================================================

func TestIntegration_Parse_PDF(t *testing.T) {
	client := newClient(t)
	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()

	result, err := client.ParsePDF(ctx, "https://arxiv.org/pdf/2301.00234.pdf", "fast")
	if err != nil {
		t.Fatalf("ParsePDF returned error: %v", err)
	}
	if !result.Success {
		t.Fatalf("ParsePDF was not successful: %+v", result.Error)
	}
	t.Logf("ParsePDF — requestId=%s mode=%s taskId=%s async=%v",
		result.RequestID, result.Mode, result.TaskID, result.Async)

	// Depending on async mode, we might get a taskId or a document
	if result.RequestID == "" && result.TaskID == "" {
		t.Error("expected either requestId or taskId to be set")
	}
}

// =======================================================================
// Group 3: IM API — Full Lifecycle
// =======================================================================

func TestIntegration_IM_FullLifecycle(t *testing.T) {
	apiClient := newClient(t)
	ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
	defer cancel()

	ts := time.Now().UnixNano()

	// ---------------------------------------------------------------
	// 3.1  Register agent A
	// ---------------------------------------------------------------
	t.Run("Account_Register_AgentA", func(t *testing.T) {})

	agentAUser := uniqueName("gotest_a")
	regResultA, err := apiClient.IM().Account.Register(ctx, &prismer.IMRegisterOptions{
		Type:         "agent",
		Username:     agentAUser,
		DisplayName:  fmt.Sprintf("Go Test Agent A %d", ts),
		AgentType:    "assistant",
		Capabilities: []string{"chat", "testing"},
		Description:  "Integration test agent A",
	})
	if err != nil {
		t.Fatalf("Register agent A error: %v", err)
	}
	if !regResultA.OK {
		t.Fatalf("Register agent A not OK: %+v", regResultA.Error)
	}

	var regDataA prismer.IMRegisterData
	if err := regResultA.Decode(&regDataA); err != nil {
		t.Fatalf("Decode register A data: %v", err)
	}
	if regDataA.Token == "" {
		t.Fatal("expected non-empty Token for agent A")
	}
	t.Logf("Agent A registered — userId=%s username=%s isNew=%v",
		regDataA.IMUserID, regDataA.Username, regDataA.IsNew)

	agentAId := regDataA.IMUserID

	// Create authenticated client for agent A
	imClientA := prismer.NewClient(regDataA.Token, prismer.WithEnvironment(prismer.Production))

	// ---------------------------------------------------------------
	// 3.2  Register agent B (target)
	// ---------------------------------------------------------------
	agentBUser := uniqueName("gotest_b")
	regResultB, err := apiClient.IM().Account.Register(ctx, &prismer.IMRegisterOptions{
		Type:         "agent",
		Username:     agentBUser,
		DisplayName:  fmt.Sprintf("Go Test Agent B %d", ts),
		AgentType:    "assistant",
		Capabilities: []string{"chat"},
		Description:  "Integration test agent B",
	})
	if err != nil {
		t.Fatalf("Register agent B error: %v", err)
	}
	if !regResultB.OK {
		t.Fatalf("Register agent B not OK: %+v", regResultB.Error)
	}

	var regDataB prismer.IMRegisterData
	if err := regResultB.Decode(&regDataB); err != nil {
		t.Fatalf("Decode register B data: %v", err)
	}
	if regDataB.Token == "" {
		t.Fatal("expected non-empty Token for agent B")
	}
	t.Logf("Agent B registered — userId=%s username=%s isNew=%v",
		regDataB.IMUserID, regDataB.Username, regDataB.IsNew)

	targetId := regDataB.IMUserID
	_ = agentAId // may be used later

	// ---------------------------------------------------------------
	// 3.3  Account.Me
	// ---------------------------------------------------------------
	t.Run("Account_Me", func(t *testing.T) {
		meResult, err := imClientA.IM().Account.Me(ctx)
		if err != nil {
			t.Fatalf("Me error: %v", err)
		}
		if !meResult.OK {
			t.Fatalf("Me not OK: %+v", meResult.Error)
		}

		var meData prismer.IMMeData
		if err := meResult.Decode(&meData); err != nil {
			t.Fatalf("Decode Me data: %v", err)
		}
		if meData.User.Username != agentAUser {
			t.Errorf("expected username=%s, got %s", agentAUser, meData.User.Username)
		}
		t.Logf("Me — user=%s role=%s stats=%+v",
			meData.User.Username, meData.User.Role, meData.Stats)
	})

	// ---------------------------------------------------------------
	// 3.4  Account.RefreshToken
	// ---------------------------------------------------------------
	t.Run("Account_RefreshToken", func(t *testing.T) {
		refreshResult, err := imClientA.IM().Account.RefreshToken(ctx)
		if err != nil {
			t.Fatalf("RefreshToken error: %v", err)
		}
		if !refreshResult.OK {
			t.Fatalf("RefreshToken not OK: %+v", refreshResult.Error)
		}

		var tokenData prismer.IMTokenData
		if err := refreshResult.Decode(&tokenData); err != nil {
			t.Fatalf("Decode token data: %v", err)
		}
		if tokenData.Token == "" {
			t.Error("expected non-empty refreshed token")
		}
		t.Logf("RefreshToken — expiresIn=%s tokenLen=%d",
			tokenData.ExpiresIn, len(tokenData.Token))
	})

	// ---------------------------------------------------------------
	// 3.5  Direct Messaging
	// ---------------------------------------------------------------
	t.Run("Direct_Send", func(t *testing.T) {
		sendResult, err := imClientA.IM().Direct.Send(ctx, targetId, "Hello from Go integration test!", nil)
		if err != nil {
			t.Fatalf("Direct.Send error: %v", err)
		}
		if !sendResult.OK {
			t.Fatalf("Direct.Send not OK: %+v", sendResult.Error)
		}
		t.Logf("Direct.Send — ok=%v", sendResult.OK)
	})

	t.Run("Direct_GetMessages", func(t *testing.T) {
		msgsResult, err := imClientA.IM().Direct.GetMessages(ctx, targetId, nil)
		if err != nil {
			t.Fatalf("Direct.GetMessages error: %v", err)
		}
		if !msgsResult.OK {
			t.Fatalf("Direct.GetMessages not OK: %+v", msgsResult.Error)
		}

		// SDK BUG: IMMessage.Metadata is map[string]any but API returns string.
		// Use imMessageLoose to work around.
		var messages []imMessageLoose
		if err := msgsResult.Decode(&messages); err != nil {
			t.Fatalf("Decode messages: %v", err)
		}
		if len(messages) == 0 {
			t.Error("expected at least one message")
		} else {
			t.Logf("Direct.GetMessages — count=%d firstContent=%q",
				len(messages), messages[0].Content)
		}
	})

	// ---------------------------------------------------------------
	// 3.6  Credits
	// ---------------------------------------------------------------
	t.Run("Credits_Get", func(t *testing.T) {
		creditsResult, err := imClientA.IM().Credits.Get(ctx)
		if err != nil {
			t.Fatalf("Credits.Get error: %v", err)
		}
		if !creditsResult.OK {
			t.Fatalf("Credits.Get not OK: %+v", creditsResult.Error)
		}

		var creditsData prismer.IMCreditsData
		if err := creditsResult.Decode(&creditsData); err != nil {
			t.Fatalf("Decode credits data: %v", err)
		}
		t.Logf("Credits.Get — balance=%.2f totalEarned=%.2f totalSpent=%.2f",
			creditsData.Balance, creditsData.TotalEarned, creditsData.TotalSpent)
	})

	t.Run("Credits_Transactions", func(t *testing.T) {
		txResult, err := imClientA.IM().Credits.Transactions(ctx, nil)
		if err != nil {
			t.Fatalf("Credits.Transactions error: %v", err)
		}
		if !txResult.OK {
			t.Fatalf("Credits.Transactions not OK: %+v", txResult.Error)
		}

		var transactions []prismer.IMTransaction
		if err := txResult.Decode(&transactions); err != nil {
			t.Fatalf("Decode transactions: %v", err)
		}
		t.Logf("Credits.Transactions — count=%d", len(transactions))
	})

	// ---------------------------------------------------------------
	// 3.7  Contacts & Discovery
	// ---------------------------------------------------------------
	t.Run("Contacts_List", func(t *testing.T) {
		contactsResult, err := imClientA.IM().Contacts.List(ctx)
		if err != nil {
			t.Fatalf("Contacts.List error: %v", err)
		}
		if !contactsResult.OK {
			t.Fatalf("Contacts.List not OK: %+v", contactsResult.Error)
		}

		var contacts []prismer.IMContact
		if err := contactsResult.Decode(&contacts); err != nil {
			t.Fatalf("Decode contacts: %v", err)
		}
		t.Logf("Contacts.List — count=%d", len(contacts))
	})

	t.Run("Contacts_Discover", func(t *testing.T) {
		discoverResult, err := imClientA.IM().Contacts.Discover(ctx, nil)
		if err != nil {
			t.Fatalf("Contacts.Discover error: %v", err)
		}
		if !discoverResult.OK {
			t.Fatalf("Contacts.Discover not OK: %+v", discoverResult.Error)
		}

		var agents []prismer.IMDiscoverAgent
		if err := discoverResult.Decode(&agents); err != nil {
			t.Fatalf("Decode discover agents: %v", err)
		}
		t.Logf("Contacts.Discover — count=%d", len(agents))
	})

	// ---------------------------------------------------------------
	// 3.8  Groups
	// ---------------------------------------------------------------
	var groupId string

	t.Run("Groups_Create", func(t *testing.T) {
		createResult, err := imClientA.IM().Groups.Create(ctx, &prismer.IMCreateGroupOptions{
			Title:       fmt.Sprintf("Go Integration Group %d", ts),
			Description: "Test group created by Go integration tests",
			Members:     []string{targetId},
		})
		if err != nil {
			t.Fatalf("Groups.Create error: %v", err)
		}
		if !createResult.OK {
			t.Fatalf("Groups.Create not OK: %+v", createResult.Error)
		}

		var groupData prismer.IMGroupData
		if err := createResult.Decode(&groupData); err != nil {
			t.Fatalf("Decode group data: %v", err)
		}
		if groupData.GroupID == "" {
			t.Fatal("expected non-empty groupId")
		}
		groupId = groupData.GroupID
		t.Logf("Groups.Create — groupId=%s title=%s members=%d",
			groupData.GroupID, groupData.Title, len(groupData.Members))
	})

	t.Run("Groups_List", func(t *testing.T) {
		if groupId == "" {
			t.Skip("no group created")
		}
		listResult, err := imClientA.IM().Groups.List(ctx)
		if err != nil {
			t.Fatalf("Groups.List error: %v", err)
		}
		if !listResult.OK {
			t.Fatalf("Groups.List not OK: %+v", listResult.Error)
		}

		var groups []prismer.IMGroupData
		if err := listResult.Decode(&groups); err != nil {
			t.Fatalf("Decode groups: %v", err)
		}
		if len(groups) == 0 {
			t.Error("expected at least one group")
		}
		t.Logf("Groups.List — count=%d", len(groups))
	})

	t.Run("Groups_Get", func(t *testing.T) {
		if groupId == "" {
			t.Skip("no group created")
		}
		getResult, err := imClientA.IM().Groups.Get(ctx, groupId)
		if err != nil {
			t.Fatalf("Groups.Get error: %v", err)
		}
		if !getResult.OK {
			t.Fatalf("Groups.Get not OK: %+v", getResult.Error)
		}

		var groupData prismer.IMGroupData
		if err := getResult.Decode(&groupData); err != nil {
			t.Fatalf("Decode group data: %v", err)
		}
		if groupData.GroupID != groupId {
			t.Errorf("expected groupId=%s, got %s", groupId, groupData.GroupID)
		}
		t.Logf("Groups.Get — groupId=%s title=%s", groupData.GroupID, groupData.Title)
	})

	t.Run("Groups_Send", func(t *testing.T) {
		if groupId == "" {
			t.Skip("no group created")
		}
		sendResult, err := imClientA.IM().Groups.Send(ctx, groupId, "Hello group from Go integration test!", nil)
		if err != nil {
			t.Fatalf("Groups.Send error: %v", err)
		}
		if !sendResult.OK {
			t.Fatalf("Groups.Send not OK: %+v", sendResult.Error)
		}
		t.Logf("Groups.Send — ok=%v", sendResult.OK)
	})

	t.Run("Groups_GetMessages", func(t *testing.T) {
		if groupId == "" {
			t.Skip("no group created")
		}
		msgsResult, err := imClientA.IM().Groups.GetMessages(ctx, groupId, nil)
		if err != nil {
			t.Fatalf("Groups.GetMessages error: %v", err)
		}
		if !msgsResult.OK {
			t.Fatalf("Groups.GetMessages not OK: %+v", msgsResult.Error)
		}

		// SDK BUG: IMMessage.Metadata is map[string]any but API returns string.
		// Use imMessageLoose to work around.
		var messages []imMessageLoose
		if err := msgsResult.Decode(&messages); err != nil {
			t.Fatalf("Decode group messages: %v", err)
		}
		if len(messages) == 0 {
			t.Error("expected at least one group message")
		} else {
			t.Logf("Groups.GetMessages — count=%d firstContent=%q",
				len(messages), messages[0].Content)
		}
	})

	// ---------------------------------------------------------------
	// 3.9  Conversations
	// ---------------------------------------------------------------
	t.Run("Conversations_List", func(t *testing.T) {
		convResult, err := imClientA.IM().Conversations.List(ctx, false, false)
		if err != nil {
			t.Fatalf("Conversations.List error: %v", err)
		}
		if !convResult.OK {
			t.Fatalf("Conversations.List not OK: %+v", convResult.Error)
		}

		var conversations []prismer.IMConversation
		if err := convResult.Decode(&conversations); err != nil {
			t.Fatalf("Decode conversations: %v", err)
		}
		t.Logf("Conversations.List — count=%d", len(conversations))
	})
}
